vsintegration/src/FSharp.Editor/AutomaticCompletion/BraceCompletionSessionProvider.fs

```diff
@@ -0,0 +1,561 @@
+            else
+
+                openingPoint <- snapshot.CreateTrackingPoint(openingSnapshotPoint.Position, PointTrackingMode.Positive)
+                let _document = snapshot.GetOpenDocumentInCurrentContextWithChanges()
+
+                if not (session.CheckOpeningPoint(this, cancellationToken)) then
+                    this.EndSession()
+                else
+
+                    use undo = this.CreateUndoTransaction()
+
+                    snapshot <-
+                        use edit = subjectBuffer.CreateEdit()
+
+                        edit.Insert(closingSnapshotPoint.Position, closingBrace.ToString()) |> ignore
+                    
+                        if edit.HasFailedChanges then
+                            Debug.Fail("Unable to insert closing brace")
+
+                            // exit without setting the closing point which will take us off the stack
+                            edit.Cancel()
+                            undo.Cancel()
+                            snapshot
+                        else
+                            edit.Apply() // FIXME: perhaps, it should be ApplyAndLogExceptions()
+
+                    let beforePoint = beforeTrackingPoint.GetPoint(textView.TextSnapshot)
+
+                    // switch from positive to negative tracking so it stays against the closing brace
+                    closingPoint <- subjectBuffer.CurrentSnapshot.CreateTrackingPoint(closingPoint.GetPoint(snapshot).Position, PointTrackingMode.Negative)
+
+                    Debug.Assert(closingPoint.GetPoint(snapshot).Position > 0 && (SnapshotSpan(closingPoint.GetPoint(snapshot).Subtract(1), 1)).GetText().Equals(closingBrace.ToString()),
+                        "The closing point does not match the closing brace character")
+
+                    // move the caret back between the braces
+                    textView.Caret.MoveTo(beforePoint) |> ignore
+
+                    session.AfterStart(this, cancellationToken)
+
+                    undo.Complete()
+
+    member __.HasNoForwardTyping(caretPoint: SnapshotPoint, endPoint: SnapshotPoint) =
+        Debug.Assert(caretPoint.Snapshot = endPoint.Snapshot, "snapshots do not match")
+
+        if caretPoint.Snapshot = endPoint.Snapshot then
+            if caretPoint = endPoint then
+                true
+            else
+                if caretPoint.Position < endPoint.Position then
+                    let span = SnapshotSpan(caretPoint, endPoint)
+                    String.IsNullOrWhiteSpace(span.GetText())
+                else
+                    false
+        else
+            false
+
+    member this.HasForwardTyping =
+        let closingSnapshotPoint = closingPoint.GetPoint(subjectBuffer.CurrentSnapshot)
+
+        if closingSnapshotPoint.Position > 0 then
+            let caretPos = getCaretPosition this
+
+            if caretPos.HasValue && not (this.HasNoForwardTyping(caretPos.Value, closingSnapshotPoint.Subtract(1))) then
```

Comment:
Isn't it better to make `getCaretPosition` return an `Option`, and match on it at this and other places?

Reply:
Yes, it would be better. This was a straight port from C# without anything special. :)